No problem, I was confused too. Basically, the code should be the following in Expr.h:

```
  // ...

  using TrailingObjects = llvm::TrailingObjects<Derived, Identifier, SourceLoc>;
  friend TrailingObjects;

  // ...

  size_t numTrailingObjects(
      typename TrailingObjects::template OverloadToken<Identifier>) const {
    return asDerived().getNumArguments();
  }

  size_t numTrailingObjects(
      typename TrailingObjects::template OverloadToken<SourceLoc>) const {
    return asDerived().hasArgumentLabelLocs() ? asDerived().getNumArguments()
                                              : 0;
  }

  // ...
```

but for some reason, MSVC gets horribly confused, and thinks these are the same functions:

> C:\Users\hughb\Documents\GitHub\swift\swift\include\swift/AST/Expr.h(681): error C2535: 'size_t swift::TrailingCallArguments<Derived>::numTrailingObjects(TrailingObjects<BaseTy,TrailingTys...>::TrailingObjectsImpl<llvm::trailing_objects_internal::AlignmentCalcHelper<TrailingTys...>::Alignment,BaseTy,llvm::TrailingObjects<BaseTy,TrailingTys...>,BaseTy,TrailingTys...>::OverloadToken) const': member function already defined or declared (compiling source file C:\Users\hughb\Documents\GitHub\swift\swift\lib\AST\ASTNode.cpp)
>  C:\Users\hughb\Documents\GitHub\swift\swift\include\swift/AST/Expr.h(675): note: see declaration of 'swift::TrailingCallArguments<Derived>::numTrailingObjects' (compiling source file C:\Users\hughb\Documents\GitHub\swift\swift\lib\AST\ASTNode.cpp)
>  C:\Users\hughb\Documents\GitHub\swift\swift\include\swift/AST/Expr.h(745): note: see reference to class template instantiation 'swift::TrailingCallArguments<Derived>' being compiled (compiling source file C:\Users\hughb\Documents\GitHub\swift\swift\lib\AST\ASTNode.cpp)

--- a/include/swift/AST/Expr.h
+++ b/include/swift/AST/Expr.h
@@ -25,7 +25,6 @@
 #include "swift/AST/TypeAlignments.h"
 #include "swift/AST/TypeLoc.h"
 #include "swift/AST/Availability.h"
-#include "swift/Basic/Compiler.h"
 #include "llvm/Support/TrailingObjects.h"
 
 namespace llvm {
@@ -659,6 +658,14 @@ public:
 template<typename Derived>
 class TrailingCallArguments
     : private llvm::TrailingObjects<Derived, Identifier, SourceLoc> {
+  // We need to friend TrailingObjects twice here to work around an MSVC bug.
+  // If we have two functions of the same name with the parameter
+  // typename TrailingObjectsIdentifier::template OverloadToken<T> where T is
+  // different for each function, then MSVC reports a "member function already
+  // defined or declared" error, which is incorrect.
+  using TrailingObjectsIdentifier = llvm::TrailingObjects<Derived, Identifier>;
+  friend TrailingObjectsIdentifier;
+
   using TrailingObjects = llvm::TrailingObjects<Derived, Identifier, SourceLoc>;
   friend TrailingObjects;
 
@@ -671,12 +678,14 @@ class TrailingCallArguments
   }
 
   size_t numTrailingObjects(
-      SWIFT_TRAILING_OBJECTS_OVERLOAD_TOKEN(Identifier)) const {
+      typename TrailingObjectsIdentifier::template OverloadToken<Identifier>)
+      const {
     return asDerived().getNumArguments();
   }
 
   size_t numTrailingObjects(
-      SWIFT_TRAILING_OBJECTS_OVERLOAD_TOKEN(SourceLoc)) const {
+      typename TrailingObjectsIdentifier::template OverloadToken<SourceLoc>)
+      const {
     return asDerived().hasArgumentLabelLocs() ? asDerived().getNumArguments()
                                               : 0;
   }

--- a/include/swift/Basic/Compiler.h
+++ b/include/swift/Basic/Compiler.h
@@ -24,17 +24,8 @@
 // https://connect.microsoft.com/VisualStudio/feedback/details/3116505
 #define SWIFT_DELETE_OPERATOR_DELETED                                          \
   { llvm_unreachable("Delete operator should not be called."); }
-
-// Work around MSVC bug: can't infer llvm::trailing_objects_internal,
-// even though we granted friend access to it.
-// https://connect.microsoft.com/VisualStudio/feedback/details/3116517
-#define SWIFT_TRAILING_OBJECTS_OVERLOAD_TOKEN(TokenType)                       \
-  llvm::trailing_objects_internal::TrailingObjectsBase::OverloadToken<TokenType>
 #else
 #define SWIFT_DELETE_OPERATOR_DELETED = delete;
-
-#define SWIFT_TRAILING_OBJECTS_OVERLOAD_TOKEN(TokenType)                       \
-  typename TrailingObjects::template OverloadToken<TokenType>
 #endif
 
 #endif // SWIFT_BASIC_COMPILER_H

--- a/include/swift/SIL/SILInstruction.h
+++ b/include/swift/SIL/SILInstruction.h
@@ -461,8 +461,8 @@ public:
     }
   }
 
-  size_t
-      numTrailingObjects(SWIFT_TRAILING_OBJECTS_OVERLOAD_TOKEN(Operand)) const {
+  size_t numTrailingObjects(
+  	  typename TrailingObjects::template OverloadToken<Operand>) const {
     return NumOperands;
   }
 
@@ -2300,8 +2300,8 @@ public:
   SILType getBoundType() const { return BoundType ; }
 
   // Implement llvm::TrailingObjects.
-  size_t
-      numTrailingObjects(SWIFT_TRAILING_OBJECTS_OVERLOAD_TOKEN(Operand)) const {
+  size_t numTrailingObjects(
+  	  typename TrailingObjects::template OverloadToken<Operand>) const {
     return NumOperands;
   }
 

